src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,960 @@
+										 workersForPlacementList);
+
+		/*
+		 * Create foreign keys if exists after the metadata changes happening in
+		 * DropShardList() and InsertSplitChildrenShardMetadata() because the foreign
+		 * key creation depends on the new metadata.
+		 */
+		CreateForeignKeyConstraints(shardGroupSplitIntervalListList,
+									workersForPlacementList);
+	}
+	PG_CATCH();
+	{
+		/* Do a best effort cleanup of shards created on workers in the above block */
+		TryDropSplitShardsOnFailure(shardGroupSplitIntervalListList,
+									workersForPlacementList);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	CitusInvalidateRelcacheByRelid(DistShardRelationId());
+}
+
+
+/* Create ShardGroup split children on a list of corresponding workers. */
+static void
+CreateSplitShardsForShardGroup(List *shardGroupSplitIntervalListList,
+							   List *workersForPlacementList)
+{
+	/* Iterate on shard interval list for shard group */
+	List *shardIntervalList = NULL;
+	foreach_ptr(shardIntervalList, shardGroupSplitIntervalListList)
+	{
+		/* Iterate on split shard interval list and corresponding placement worker */
+		ShardInterval *shardInterval = NULL;
+		WorkerNode *workerPlacementNode = NULL;
+		forboth_ptr(shardInterval, shardIntervalList, workerPlacementNode,
+					workersForPlacementList)
+		{
+			/* Populate list of commands necessary to create shard interval on destination */
+			List *splitShardCreationCommandList = GetPreLoadTableCreationCommands(
+				shardInterval->relationId,
+				false, /* includeSequenceDefaults */
+				NULL /* auto add columnar options for cstore tables */);
+			splitShardCreationCommandList = WorkerApplyShardDDLCommandList(
+				splitShardCreationCommandList,
+				shardInterval->shardId);
+
+			/* Create new split child shard on the specified placement list */
+			CreateObjectOnPlacement(splitShardCreationCommandList, workerPlacementNode);
+		}
+	}
+}
+
+
+/* Create ShardGroup auxiliary structures (indexes, stats, replicaindentities, triggers)
+ * on a list of corresponding workers.
+ */
+static void
+CreateAuxiliaryStructuresForShardGroup(List *shardGroupSplitIntervalListList,
+									   List *workersForPlacementList)
+{
+	/*
+	 * Create auxiliary structures post copy.
+	 */
+	List *shardIntervalList = NULL;
+	foreach_ptr(shardIntervalList, shardGroupSplitIntervalListList)
+	{
+		/* Iterate on split shard interval list and corresponding placement worker */
+		ShardInterval *shardInterval = NULL;
+		WorkerNode *workerPlacementNode = NULL;
+		forboth_ptr(shardInterval, shardIntervalList, workerPlacementNode,
+					workersForPlacementList)
+		{
+			List *indexCommandList = GetPostLoadTableCreationCommands(
+				shardInterval->relationId,
+				true /* includeIndexes */,
+				true /* includeReplicaIdentity */);
+			indexCommandList = WorkerApplyShardDDLCommandList(
+				indexCommandList,
+				shardInterval->shardId);
+
+			CreateObjectOnPlacement(indexCommandList, workerPlacementNode);
```

Comment:
I'm kind of wondering, we run into serious issues with the number of tables sometimes, and we could end up with quite a staggering number of objects. Would it be possible to capture this in a ExecuteTaskListOutsideTransaction?

Reply:
Updated.

Reply:
Fixed.